src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -691,72 +693,130 @@ putPaddingTo64 written
+    let getPositive off64 = do
+          let w = indexByteArray ba off64 :: Word64
+          when (w > fromIntegral (maxBound :: Int)) $
+            Left "Size information is too large for Int"
+          return (fromIntegral w)
+
+    rfprec <- getPositive (len64 - 3)
+    numPages <- getPositive (len64 - 2)
+    numEntries <- getPositive (len64 - 1)
+
+    when (rfprec > snd rangeFinderPrecisionBounds) $
+      Left "Invalid range finder precision"
+    let numRanges = 2 ^ rfprec + 1
 
+    -- read vectors
     -- offsets in 32 bits
-    let offset1_32 = 1  -- after version indicator
-    let (offset2_32, ciPrimary) = getVec32 ba offset1_32 numPages
-    let (offset3_32, ciRangeFinder) = getVec32 ba offset2_32 (2 ^ rfprec + 1)
+    let off1_32 = 1  -- after version indicator
+    (!off2_32, ciPrimary) <- getVec32 "Primary array" ba off1_32 numPages
+    (!off3_32, ciRangeFinder) <- getVec32 "Range finder" ba off2_32 numRanges
     -- offsets in 64 bits
-    let offset3 = ceilDiv2 offset3_32
-    let (offset4, ciClashes) = getBitVec ba offset3 numPages
-    let (offset5, ciLargerThanPage) = getBitVec ba offset4 numPages
-    let ciTieBreaker = getTieBreaker ba offset5
+    let !off3 = ceilDiv2 off3_32
+    (!off4, ciClashes) <- getBitVec "Clash bit vector" ba off3 numPages
+    (!off5, ciLargerThanPage) <- getBitVec "LTP bit vector" ba off4 numPages
+    (!off6, ciTieBreaker) <- getTieBreaker ba off5
+
+    let bytesUsed = mul8 (off6 + 3)
+    when (bytesUsed > sizeofByteArray ba) $
+      Left "Byte array is too small for components"
+    when (bytesUsed < sizeofByteArray ba) $
+      Left "Byte array is too large for components"
 
     let ciRangeFinderPrecision = rfprec
     return (NumEntries numEntries, CompactIndex {..})
 
 type Offset32 = Int
 type Offset64 = Int
 
-getVec32 :: ByteArray -> Offset32 -> Int -> (Offset32, VU.Vector Word32)
-getVec32 ba offset32 numEntries = (offset32 + numEntries, vec)
-  where
-    vec = VU.V_Word32 (VP.Vector offset32 numEntries ba)
-
-getBitVec :: ByteArray -> Offset64 -> Int -> (Offset64, VU.Vector Bit)
-getBitVec ba offset64 numEntries = (offset64', vec)
+getVec32 ::
+     String -> ByteArray -> Offset32 -> Int
+  -> Either String (Offset32, VU.Vector Word32)
+getVec32 name ba off32 numEntries =
+    case checkedPrimVec off32 numEntries ba of
+      Nothing  -> Left (name <> " is out of bounds")
+      Just vec -> Right (off32 + numEntries, VU.V_Word32 vec)
+
+getBitVec ::
+     String -> ByteArray -> Offset64 -> Int
+  -> Either String (Offset64, VU.Vector Bit)
+getBitVec name ba off numEntries =
+    case checkedBitVec (mul64 off) numEntries ba of
+      Nothing  -> Left (name <> " is out of bounds")
+      Just vec -> Right (off + ceilDiv64 numEntries, vec)
+
+-- | Checks bounds.
+--
+-- Inefficient, but okay for a small number of entries.
+getTieBreaker ::
+     ByteArray -> Offset64
+  -> Either String (Offset64, Map SerialisedKey PageNo)
+getTieBreaker ba = \off -> do
```

Comment:
The offsets in this function aren't converted from `Word64` to `Word8` offsets

Reply:
Yes, it should be `mul 8 off >= sizeofByteArray ba`. Good catch!